pkg/controller/instance/garbage_collection.go

```diff
@@ -0,0 +1,106 @@
+package instance
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	h "github.com/mittwald/goharbor-client"
+	registriesv1alpha1 "github.com/mittwald/harbor-operator/pkg/apis/registries/v1alpha1"
+	"github.com/mittwald/harbor-operator/pkg/controller/internal"
+)
+
+func (r *ReconcileInstance) reconcileGarbageCollection(ctx context.Context, harbor *registriesv1alpha1.Instance) error {
+	// Build a client to connect to the harbor API
+	harborClient, err := internal.BuildClient(ctx, r.client, harbor)
+	if err != nil {
+		return err
+	}
+
+	// Get the held garbage collection (Harbor AdminJob) from the Harbor API
+	// While the API returns HTTP status 200 if garbage collection is not yet configured,
+	// the object received is an empty h.AdminJobReq{}
+	gc, err := harborClient.System().GetSystemGarbageCollectionSchedule()
+	if err != nil {
+		return err
+	}
+
+	// Construct a new garbage collection from the provided instance spec
+	newGc, err := BuildGarbageCollectionScheduleFromSpec(harbor)
+	if err != nil {
+		return err
+	}
+
+	// Assume that the garbage collection's Schedule is nil, if it has not been created yet
+	if gc.AdminJobSchedule == nil {
+		err = harborClient.System().CreateSystemGarbageCollectionSchedule(newGc)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Compare the constructed garbage collection to the existing one and update accordingly
+	if !reflect.DeepEqual(newGc, gc) {
+		err = harborClient.System().UpdateSystemGarbageCollectionSchedule(newGc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// BuildGarbageCollectionScheduleFromSpec constructs and returns a Harbor AdminJobReq from the spec
+func BuildGarbageCollectionScheduleFromSpec(harbor *registriesv1alpha1.Instance) (h.AdminJobReq, error) {
+	if harbor.Spec.GarbageCollection.Schedule == nil {
```

Comment:
Maybe also test for `harbor.Spec.GarbageCollection == nil`? Or is that already done someplace else?

Reply:
OK, I noticed that that condition is already checked in `ReconcileInstance.Reconcile()`; but this method is public (does it need to be?), so there may be other code paths leading into this function (now or in the future).